src/smt/theory_str_noodler/nielsen_decision_procedure.cpp

```diff
@@ -0,0 +1,567 @@
+     */
+    Formula NielsenDecisionProcedure::trim_formula(const Formula& formula) const {
+        Formula ret;
+
+        for(const Predicate& pred : formula.get_predicates()) {
+            auto params = pred.get_params();
+            size_t len = std::min(params[0].size(), params[1].size());
+            size_t i = 0;
+            for(; i < len; i++) {
+                if(params[0][i] != params[1][i]) {
+                    break;
+                }
+            }
+            std::vector<Concat> sides({
+                Concat(params[0].begin()+i, params[0].end()),
+                Concat(params[1].begin()+i, params[1].end())
+            });
+            ret.add_predicate(Predicate(PredicateType::Equation, sides));
+        }
+        return ret;
+    }
+
+    /**
+     * @brief Check if a predicate is trivially unsat.
+     * 
+     * @param pred Predicate
+     * @return true -> unsat
+     */
+    bool NielsenDecisionProcedure::is_pred_unsat(const Predicate& pred) const {
+        Concat left = pred.get_left_side();
+        Concat right = pred.get_right_side();
+
+        if(left.size() == 0 && right.size() > 0 && right[0].is_literal()) {
+            return true;
+        }
+        if(right.size() == 0 && left.size() > 0 && left[0].is_literal()) {
+            return true;
+        }
+        if(left.size() > 0 && right.size() > 0 && right[0].is_literal() && left[0].is_literal() && right[0] != left[0]) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @brief Create a counter system from the Nielsen graph.
+     * 
+     * @param graph Graph to be converted to the counter system.
+     * @return CounterSystem
+     */
+    CounterSystem NielsenDecisionProcedure::create_counter_system(const NielsenGraph& graph) const {
+        CounterSystem ret;
+
+        // conversion of a nielsen label to the counter label
+        auto conv_fnc = [&](const NielsenLabel& lab) {
+            if(lab.second.size() == 0) {
+                return CounterLabel{lab.first, {BasicTerm(BasicTermType::Length, "0")}};
+            } else if(lab.second.size() == 2 && lab.second[0].is_literal()) {
+                return CounterLabel{lab.first, {lab.second[1], BasicTerm(BasicTermType::Length, "1")}};
+            } else if(lab.second.size() == 2 && lab.second[0].is_variable()) {
+                return CounterLabel{lab.first, {lab.second[1], lab.second[0]}};
+            } else {
+                util::throw_error("create_counter_system: unsupported label type");
+            }
+        };
+
+        // switch initial and final nodes
+        for(const Formula& fin : graph.get_fins()) {
+            ret.set_init(fin);
+        }
+        ret.add_fin(graph.get_init());
+        // reverse edges
+        for(const auto& pr : graph.edges) {
+            for(const auto& trans : pr.second) {
+                ret.add_edge(trans.first, pr.first, conv_fnc(trans.second));
+            }
+        }
+        return ret;
```

Comment:
We have a function for this already, no?

Reply:
Not really. In this case we want to apply the lambda on each label.